File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.io.*;
import java.util.*;

class Solution {
    boolean solution(String s) {
        
        boolean answer = true;
        
        int lengthOfs = s.length(); // 받은 문자열의 길이
        Character[] array = new Character[lengthOfs];
        
        Stack<Character> stack = new Stack<>();
        
        for(int i = 0; i < lengthOfs; i++) {
            array[i] = s.charAt(i);
        } // 일단 배열에 하나씩 넣고
                
        for(int i = 0; i < lengthOfs; i++) {
            
            if(stack.isEmpty()) {
                if(array[i] == ')')  {
                    return false;
                }
                stack.push(array[i]);
                continue;
            } // 스택이 비어있으면 일단 첫번째 원소 넣고
            
            // 스택에 뭐라도 있으면(2번쨰 원소 확인부터)
            Character c = stack.peek();
            if(array[i] == c) {
                stack.push(c);
            }
            else stack.pop();            
        }
        
        if(!stack.isEmpty()) answer = false;
        return answer;
        
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

bool solution(string s){
    
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    
    bool answer = true;
    
    // (가 먼저 나올때랑, )가 먼저 나올때 구분해서 생각해보면
    // 일단 (가 나오면 뒤도 안 돌아보고 false 
    // 그런데 )가 나오면 생각 좀 해봐야함
    
    stack<char> stack;
    
    for(int i = 0; i < s.size(); i++) {

        if(s[i] == '(') {
            stack.push(s[i]); // 일단 (가 나오면 저장
        } else {
            if(!stack.empty()) {
                stack.pop();
            } else return false;
        }    
    }
    
    if(!stack.empty()) answer = false;
    return answer;
    
}